Mathlib/MeasureTheory/Measure/LevyProkhorovMetric.lean

```diff
@@ -395,13 +444,188 @@ lemma continuous_levyProkhorov_to_probabilityMeasure :
                ge_iff_le, forall_exists_index]
     refine ⟨0, fun a i hia ↦ le_trans (integral_nonneg f_nn) (hia i rfl.le)⟩
 
-/-- The topology of the Lévy-Prokhorov metric is finer than the topology of convergence in
+/-- The topology of the Lévy-Prokhorov metric is at least as fine as the topology of convergence in
 distribution. -/
 theorem levyProkhorov_le_convergenceInDistribution :
     TopologicalSpace.coinduced (LevyProkhorov.probabilityMeasure (Ω := Ω)) inferInstance
       ≤ (inferInstance : TopologicalSpace (ProbabilityMeasure Ω)) :=
   (continuous_levyProkhorov_to_probabilityMeasure).coinduced_le
 
-end Levy_Prokhorov_comparison
+end Levy_Prokhorov_is_finer
+
+
+
+section Levy_Prokhorov_metrizes_convergence_in_distribution
+
+/-! ### On separable spaces the Lévy-Prokhorov distance metrizes convergence in distribution -/
+
+open BoundedContinuousFunction TopologicalSpace
+
+variable {ι : Type*} {Ω : Type*} [MeasurableSpace Ω]
+
+/-- In a separable pseudometric space, for any ε > 0 there exists a countable collection of
+disjoint Borel measurable subsets of diameter at most ε that cover the whole space. -/
+lemma SeparableSpace.exists_measurable_partition_diam_le
+    (X : Type*) [PseudoMetricSpace X] [SeparableSpace X]
+    [MeasurableSpace X] [OpensMeasurableSpace X] {ε : ℝ} (ε_pos : 0 < ε) :
+    ∃ (As : ℕ → Set X), (∀ n, MeasurableSet (As n)) ∧ (∀ n, Bornology.IsBounded (As n)) ∧
```

Comment:
Is there a reason you're including the boundedness condition here, while it also follows from the next condition on the diameter?

Reply:
Ah, I had overlooked that! Maybe add a comment somewhere about that? Depending on your application, I'll let you judge if it is more efficient to switch to `ediam` or keep the current version.

Reply:
I struggled a little trying to do the `EMetric.diam` version directly, probably due to my lack of familiarity with uniformities and possibly due to some missing API on emetric and uniformities (in particular I failed to find the analogue of `Metric.ball_eq_ball'` to close the remaining sorry below).

```
/-- In a separable pseudoemetric space, for any ε > 0 there exists a countable collection of
disjoint Borel measurable subsets of diameter at most ε that cover the whole space. -/
lemma SeparableSpace.exists_measurable_partition_ediam_le
    {X : Type*} [PseudoEMetricSpace X] [MeasurableSpace X] [OpensMeasurableSpace X]
    [SeparableSpace X] {ε : ℝ≥0∞} (ε_pos : 0 < ε) :
    ∃ (As : ℕ → Set X), (∀ n, MeasurableSet (As n)) ∧
        (∀ n, EMetric.diam (As n) ≤ ε) ∧ (⋃ n, As n = univ) ∧
        (Pairwise (fun (n m : ℕ) ↦ Disjoint (As n) (As m))) := by
  by_cases X_emp : IsEmpty X
  · refine ⟨fun _ ↦ ∅, fun _ ↦ MeasurableSet.empty, ?_, ?_, fun _ _ _ ↦ disjoint_of_subsingleton⟩
    · intro n
      simpa only [EMetric.diam_empty] using LT.lt.le ε_pos
    · simp only [iUnion_empty]
      apply Eq.symm
      simp only [univ_eq_empty_iff, X_emp]
  rw [not_isEmpty_iff] at X_emp
  obtain ⟨xs, xs_dense⟩ := exists_dense_seq X
  have half_ε_pos : 0 < ε / 2 := ENNReal.div_pos ε_pos.ne.symm two_ne_top
  set Bs := fun n ↦ EMetric.ball (xs n) (ε/2)
  set As := disjointed Bs
  set Bs := fun n ↦ EMetric.ball (xs n) (ε/2)
  set As := disjointed Bs
  refine ⟨As, ?_, ?_, ?_, ?_⟩
  · exact MeasurableSet.disjointed (fun n ↦ measurableSet_eball)
  · intro n
    apply (EMetric.diam_mono (disjointed_subset Bs n)).trans
    convert EMetric.diam_ball
    exact (ENNReal.mul_div_cancel' two_ne_zero two_ne_top).symm
  · have aux : ⋃ n, Bs n = univ := by
      convert DenseRange.iUnion_uniformity_ball xs_dense <| edist_mem_uniformity half_ε_pos
      simp only [Bs]
      -- ⊢ EMetric.ball (xs x✝) (ε / 2) = UniformSpace.ball (xs x✝) {p | edist p.1 p.2 < ε / 2}
      sorry
    simpa only [← aux] using iUnion_disjointed
  · exact disjoint_disjointed Bs
```

Reply:
It is of course possible to deduce the `EMetric.diam` version from the version I stated originally (below). But at that point it is in my opinion already dubious whether the pros of `EMetric.diam` outweigh the extra hassle.

```

```
lemma ENNReal.truncateToReal_pos (z t : ℝ≥0∞) (z_pos : 0 < z) (t_pos : 0 < t) (h : z ≠ ∞ ∨ t ≠ ∞) :
    0 < z.truncateToReal t := by
  simp [ENNReal.truncateToReal]
  apply toReal_pos (lt_min z_pos t_pos).ne.symm
  rcases h with z_ne_top | t_ne_top
  · exact (lt_of_le_of_lt (min_le_left z t) z_ne_top.symm.lt_top').ne
  · exact (lt_of_le_of_lt (min_le_right z t) t_ne_top.symm.lt_top').ne

/-- In a separable pseudometric space, for any ε > 0 there exists a countable collection of
disjoint Borel measurable subsets of diameter at most ε that cover the whole space. -/
lemma SeparableSpace.exists_measurable_partition_ediam_le {ε : ℝ≥0∞} (ε_pos : 0 < ε) :
    ∃ (As : ℕ → Set Ω), (∀ n, MeasurableSet (As n)) ∧
        (∀ n, EMetric.diam (As n) ≤ ε) ∧ (⋃ n, As n = univ) ∧
        (Pairwise (fun (n m : ℕ) ↦ Disjoint (As n) (As m))) := by
  obtain ⟨As, As_mble, As_bdd, As_diam_le, As_cover, As_disj⟩ :=
    SeparableSpace.exists_measurable_partition_diam_le Ω
      <| ENNReal.truncateToReal_pos _ _ ε_pos zero_lt_one (Or.inr one_ne_top)
  refine ⟨As, As_mble, ?_, As_cover, As_disj⟩
  by_cases ε_top : ε = ∞
  · simp [ε_top]
  exact fun n ↦ (toReal_le_toReal (isBounded_iff_ediam_ne_top.mp (As_bdd n)) ε_top).mp
    <| le_trans (As_diam_le n) (truncateToReal_le ε_top)
```
```

Reply:
I think I am inclined to keep the original spelling for now.